### PR TITLE
Features - Building Blocks Enumerables - TSE Review

### PR DESCRIPTION
**Pull request by fdfragoso**
Module Enumerable implemented;
Methods implemented:
  - my_each
  - my_each_with_index
  - my_select
  - my_all?
  - my_any?
  - my_none? 
  - my_count
  - my_map
  - my_inject
  - multiply_els

Unit tests made!
Linters checked!

### CODE REVIEW

File: enumerables.rb [hunk 1]
```diff
@@ -0,0 +1,143 @@
+module Enumerable
+  def my_each
```
Comment (Forison) [inline enumerables.rb:3]: * [x] There should be a working my_each identical and returns the same thing as ruby's [each](https://ruby-doc.org/core-2.6.4/Array.html#method-i-each). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_each.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))
#### Let us compare the results return in each of the examples below and make the return the same result as the original method.
1. ```p [1,2,3,4,5].each{|x|  x}``` and ```p [1,2,3,4,5].my_each{|x|  x}``` must return same result.
2. ```p [1,2,3,4,5].each ``` and ```p [1,2,3,4,5].my_each``` must return same result.
3. ```p (0..5).each{|x|  x}``` and ```p (0..5).my_each{|x|  x}``` must return same result.
3. ```p (0...5).each{|x|  x}``` and ```(0...5).my_each{|x|  p x}``` must return same result.

File: enumerables.rb [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+    i = 0
+    while i < length
+      yield (self[i])
+      i += 1
+    end
+  end
+
+  def my_each_with_index
```
Comment (Forison) [inline enumerables.rb:14]: * [x] There should be working `my_each_with_index`  identical and returns the same thing as ruby's [`each_with_index`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-each_with_index). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_each_with_index.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))
#### Let us compare the results return in each of the examples below and make the return the same result as the original method.
1. ```p [1,2,3,4,5].each_with_index``` and ```p [1,2,3,4,5].my_each_with_index``` must return same result.
2. ```p [1,2,3,4,5].each_with_index{|a| a}``` and ```p [1,2,3,4,5].my_each_with_index{|a| a}``` must return same result.
3. ```(0..5).each_with_index{|a,b| p b}``` and ```(0..5).my_each_with_index{|a,b| p b}``` must return same result.

File: enumerables.rb [hunk 3]
```diff
@@ -0,0 +1,143 @@
...
+    i = 0
+    while i < length
+      yield self[i], i
+      i += 1
+    end
+  end
+
+  def my_select
```
Comment (Forison) [inline enumerables.rb:25]: * [x] There should be a working `my_select`  identical and returns the same thing as ruby's [`select`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-select). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_select.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))
#### Let us compare the results return in each of the examples below and make the return the same result as the original method.
1. ```p [1,2,3,4,5].select``` and ```p [1,2,3,4,5].my_select``` must return same result.
2. ```p (0..5).select{|x| x.even?} ``` and ```p (0..5).my_select{|x| x.even?}``` must return same result.

File: enumerables.rb [hunk 4]
```diff
@@ -0,0 +1,143 @@
...
+    selected = []
+    my_each { |i| selected << i if yield(i) }
+    selected
+  end
+
+  def my_all?
+    result = true
```
Comment (Forison) [inline enumerables.rb:37]: * [x] There should be a working `my_all?`  identical and returns the same thing as ruby's [`all?`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-all-3F). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_all.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))
#### Let us compare the results return in each of the examples below and make the return the same result as the original method.
1. ```p %w[ant bear cat].all?(/t/)``` and ```p %w[ant bear cat].my_all?(/t/)``` must return same result.
2. ```p [1, 2i, 3.14].all?(Numeric)``` and ```p [1, 2i, 3.14].my_all?(Numeric)``` must return same result.
2. ```p [nil, true, 99].all?``` and ```p [nil, true, 99].my_all?``` must return same result.

File: enumerables.rb [hunk 5]
```diff
@@ -0,0 +1,143 @@
...
+    my_each { |i| result = false unless yield(i) }
+    result
+  end
+
+  def my_any?
+    result = false
```
Comment (Forison) [inline enumerables.rb:51]: * [x] There should be working `my_any?` identical and returns the same thing as ruby's [`any?`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-any-3F). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_any.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))
#### Let us compare the results return in each of the examples below and make the return the same result as the original method.
1. ```p %w[ant bear cat].any?(/d/)``` and ```p %w[ant bear cat].my_any?(/d/)``` must return same result.
2. ```p [nil, true, 99].any?(Integer)``` and ```p [nil, true, 99].my_any?(Integer)``` must return same result.
3. ```p [nil, true, 99].any? ``` and ```p [nil, true, 99].my_any? ``` must return same result.

File: enumerables.rb [hunk 6]
```diff
@@ -0,0 +1,143 @@
...
+    my_each { |i| result = true if yield(i) }
+    result
+  end
+
+  def my_none?
+    result = true
```
Comment (Forison) [inline enumerables.rb:65]: * [x] There should be a working `my_none?`  identical and returns the same thing as ruby's [`none?`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-none-3F). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_none.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))
#### Let us compare the results return in each of the examples below and make the return the same result as the original method.
1. ```p %w{ant bear cat}none?(/d/)``` and ```p %w{ant bear cat}.my_none?(/d/)``` must return same result.
1. ```p [1, 3.14, 42].none?(Float)``` and ```p [1, 3.14, 42].my_none?(Float)``` must return same result.
1. ```p [nil].none?``` and ```p [nil].my_none?``` must return same result.
1. ```p [nil, false].none?``` and ```p [nil, false].my_none?``` must return same result.
1. ```p [nil, false, true].none?``` and ```p [nil, false, true].my_none?``` must return same result.

File: enumerables.rb [hunk 7]
```diff
@@ -0,0 +1,143 @@
...
+    my_each { |i| result = false if yield(i) }
+    result
+  end
+
+  def my_count
+    result = 0
```
Comment (Forison) [inline enumerables.rb:80]: * [ ]There should be a working `my_count`  identical and returns the same thing as ruby's [`count`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-count). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_count.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))
#### Let us compare the results return in each of the examples below and make the return the same result as the original method.
1. ``` p [1, 2, 4, 2].count ``` and ```p [1, 2, 4, 2].my_count``` must return same result.
2. ``` p [1, 2, 4, 2].count(2)  ``` and p ``` p [1, 2, 4, 2].my_count(2)  ``` must return same result.

File: enumerables.rb [hunk 8]
```diff
@@ -0,0 +1,143 @@
...
+    my_each { |i| result += 1 if yield(i) }
+    result
+  end
+
+  def my_map(arr = nil)
+    result = []
```
Comment (Forison) [inline enumerables.rb]: * [x] There should be a working `my_map`  identical and returns the same thing as ruby's [`map`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-map). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_map.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))
#### Let us compare the results return in each of the examples below and make the return the same result as the original method.
```p (1..4).map { |i| i*i }``` and ```p (1..4).my_map { |i| i*i }``` must return same result.
There is also an extra functionality for my_map. It should be able to accept a ```Proc```.

File: enumerables.rb [hunk 9]
```diff
@@ -0,0 +1,143 @@
...
+    if arr.nil?
+      my_each { |i| result.push(yield(i)) }
+    else
+      my_each { |i| result.push result.call(i) }
+    end
+    result
+  end
+
+  def my_inject
+    result = first
```
Comment (Forison) [inline enumerables.rb]: * [x] There should be a working `my_inject`  identical and returns the same thing as ruby's [`inject`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-inject). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_inject.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))
#### Let us compare the results return in each of the examples below and make the return the same result as the original method.
```p (5..10).inject(1) { |product, n| product * n }``` and ```p (5..10).my_inject(1) { |product, n| product * n }``` must return same result.

File: enumerables.rb [hunk 10]
```diff
@@ -0,0 +1,143 @@
...
+    each_with_index do |item, index|
+      next if index.zero?
+
+      result = yield(result, item)
+    end
+    result
+  end
+
+  def multiply_els
```
Comment (Forison) [inline enumerables.rb]: * [x] ```multiply_els``` this is not an enumerable method, kindly move it outside the module wrap as required.

**Review by Forison — CHANGES_REQUESTED** [review]
## Status:: Changes Required ❌
# 🙋‍♂️ Team!!!
#### You have done a great job so far, however, there a few changes we have to make to get this project approved👨‍🔧.
* [x] Kindly make remove the test examples from the ```enumerable.rb``` file.
* [x] Check the comments 👇below👇 for the changes to be effected to have the project approved.
## That is it, kindly make the changes and submit the project for a re-review🤗.
#### Happy coding
Addo Forison

**Comment by fdfragoso** [comment]
Some rubocop errors, but all due to ===. I checked on slack that this is ok for this project.

File: enumerables.rb [hunk 11]
```diff
@@ -0,0 +1,122 @@
...
+          0
+        else
+          var
+        end
+
+    while i < size
+      yield(to_a[i])
+      i += 1
+    end
+  end
```
Comment (maelfosso) [inline enumerables.rb]:
```rb
2.6.5 :009 > array
 => [0, 7, 1, 5, 0, 3, 3, 2, 1, 2, 5, 2, 5, 6, 2, 1, 8, 4, 7, 2, 7, 5, 7, 2, 7, 1, 8, 5, 5, 2, 8, 8, 6, 7, 3, 2, 1, 2, 4, 3, 3, 7, 6, 1, 0, 4, 3, 3, 1, 0, 7, 2, 2, 8, 1, 3, 2, 2, 3, 5, 0, 4, 1, 0, 3, 6, 8, 1, 4, 0, 7, 8, 7, 0, 8, 1, 3, 7, 4, 2, 8, 8, 5, 5, 3, 7, 6, 0, 1, 5, 7, 1, 8, 6, 2, 6, 4, 5, 1, 6] 
2.6.5 :010 > block = proc { |num| num < 4 }
 => #<Proc:0x0000000000d15088@(irb):10> 
2.6.5 :011 > array.my_each(&block)
 => nil 
2.6.5 :012 > array.each(&block)
 => [0, 7, 1, 5, 0, 3, 3, 2, 1, 2, 5, 2, 5, 6, 2, 1, 8, 4, 7, 2, 7, 5, 7, 2, 7, 1, 8, 5, 5, 2, 8, 8, 6, 7, 3, 2, 1, 2, 4, 3, 3, 7, 6, 1, 0, 4, 3, 3, 1, 0, 7, 2, 2, 8, 1, 3, 2, 2, 3, 5, 0, 4, 1, 0, 3, 6, 8, 1, 4, 0, 7, 8, 7, 0, 8, 1, 3, 7, 4, 2, 8, 8, 5, 5, 3, 7, 6, 0, 1, 5, 7, 1, 8, 6, 2, 6, 4, 5, 1, 6] 
2.6.5 :013 > 
```

File: enumerables.rb [hunk 12]
```diff
@@ -0,0 +1,122 @@
...
+
+  def my_each_with_index
+    return to_enum unless block_given?
+
+    i = 0
+    while i < size
+      yield(to_a[i], i)
+      i += 1
+    end
+  end
```
Comment (maelfosso) [inline enumerables.rb]:
```rb
2.6.5 :015 > array
 => [0, 7, 1, 5, 0, 3, 3, 2, 1, 2, 5, 2, 5, 6, 2, 1, 8, 4, 7, 2, 7, 5, 7, 2, 7, 1, 8, 5, 5, 2, 8, 8, 6, 7, 3, 2, 1, 2, 4, 3, 3, 7, 6, 1, 0, 4, 3, 3, 1, 0, 7, 2, 2, 8, 1, 3, 2, 2, 3, 5, 0, 4, 1, 0, 3, 6, 8, 1, 4, 0, 7, 8, 7, 0, 8, 1, 3, 7, 4, 2, 8, 8, 5, 5, 3, 7, 6, 0, 1, 5, 7, 1, 8, 6, 2, 6, 4, 5, 1, 6] 
2.6.5 :016 > block = proc { |num| num < 4 }
 => #<Proc:0x0000000000e612e8@(irb):16> 
2.6.5 :017 > array.my_each_with_index(&block)
 => nil 
2.6.5 :018 > array.each_with_index(&block)
 => [0, 7, 1, 5, 0, 3, 3, 2, 1, 2, 5, 2, 5, 6, 2, 1, 8, 4, 7, 2, 7, 5, 7, 2, 7, 1, 8, 5, 5, 2, 8, 8, 6, 7, 3, 2, 1, 2, 4, 3, 3, 7, 6, 1, 0, 4, 3, 3, 1, 0, 7, 2, 2, 8, 1, 3, 2, 2, 3, 5, 0, 4, 1, 0, 3, 6, 8, 1, 4, 0, 7, 8, 7, 0, 8, 1, 3, 7, 4, 2, 8, 8, 5, 5, 3, 7, 6, 0, 1, 5, 7, 1, 8, 6, 2, 6, 4, 5, 1, 6] 
2.6.5 :019 > 
```

File: enumerables.rb [hunk 13]
```diff
@@ -0,0 +1,122 @@
...
+      if block_given?
+        return result = true if yield(item)
+      elsif var.nil?
+        return result = true if item
+      else
+        return result = true if var === item
+      end
+    end
+    result
+  end
```
Comment (maelfosso) [inline enumerables.rb]:
```rb
2.6.5 :025 > false_block = proc { |num| num > 9 }
 => #<Proc:0x0000000000eb7be8@(irb):25> 
2.6.5 :026 > array.my_any?(&false_block)
 => true 
2.6.5 :027 > array.any?(&false_block)
 => false 
2.6.5 :028 >
```

Comment (maelfosso) [reply enumerables.rb]:
```rb
2.6.5 :032 > false_array = [nil, false, nil, false]
 => [nil, false, nil, false] 
2.6.5 :033 > false_array.any?
 => false 
2.6.5 :034 > false_array.my_any?
 => true 
2.6.5 :035 > 
```

Comment (maelfosso) [reply enumerables.rb]:
```rb
2.6.5 :042 > words = %w[dog door rod blade]
 => ["dog", "door", "rod", "blade"] 
2.6.5 :043 > words.any?(Integer)
 => false 
2.6.5 :044 > words.my_any?(Integer)
 => true 
2.6.5 :045 >
```

Comment (maelfosso) [reply enumerables.rb]:
```rb
2.6.5 :047 > words
 => ["dog", "door", "rod", "blade"] 
2.6.5 :048 > words.any?(/z/)
 => false 
2.6.5 :049 > words.my_any?(/z/)
 => true 
2.6.5 :050 >
```

Comment (maelfosso) [reply enumerables.rb]:
```rb
2.6.5 :052 > words
 => ["dog", "door", "rod", "blade"] 
2.6.5 :053 > words.any?('cat')
 => false 
2.6.5 :054 > words.my_any?('cat')
 => true 
2.6.5 :055 >
```

File: enumerables.rb [hunk 14]
```diff
@@ -0,0 +1,122 @@
...
+  def my_inject(*args)
+    if args.count.zero?
+      my_each { |num| args = yield(args, num) }
+    else
+      args = args[0]
+      my_each { |num| args = yield(args, num) }
+      args
+    end
+  end
+end
```
Comment (maelfosso) [inline enumerables.rb]:
```rb
2.6.5 :057 > operation = proc { |sum, n| sum + n }
 => #<Proc:0x0000000000f782f8@(irb):57> 
2.6.5 :058 > array
 => [0, 7, 1, 5, 0, 3, 3, 2, 1, 2, 5, 2, 5, 6, 2, 1, 8, 4, 7, 2, 7, 5, 7, 2, 7, 1, 8, 5, 5, 2, 8, 8, 6, 7, 3, 2, 1, 2, 4, 3, 3, 7, 6, 1, 0, 4, 3, 3, 1, 0, 7, 2, 2, 8, 1, 3, 2, 2, 3, 5, 0, 4, 1, 0, 3, 6, 8, 1, 4, 0, 7, 8, 7, 0, 8, 1, 3, 7, 4, 2, 8, 8, 5, 5, 3, 7, 6, 0, 1, 5, 7, 1, 8, 6, 2, 6, 4, 5, 1, 6] 
2.6.5 :059 > array.my_inject(&operation)
Traceback (most recent call last):
        9: from /usr/share/rvm/rubies/ruby-2.6.5/bin/irb:23:in `<main>'
        8: from /usr/share/rvm/rubies/ruby-2.6.5/bin/irb:23:in `load'
        7: from /usr/share/rvm/rubies/ruby-2.6.5/lib/ruby/gems/2.6.0/gems/irb-1.0.0/exe/irb:11:in `<top (required)>'
        6: from (irb):59
        5: from enumerables.rb:111:in `my_inject'
        4: from enumerables.rb:12:in `my_each'
        3: from enumerables.rb:111:in `block in my_inject'
        2: from (irb):57:in `block in irb_binding'
        1: from (irb):57:in `+'
TypeError (no implicit conversion of Integer into Array)
2.6.5 :060 > array.inject(&operation)
 => 390 
2.6.5 :061 > 
```

Comment (maelfosso) [reply enumerables.rb]:
```rb
2.6.5 :069 > array
 => [0, 7, 1, 5, 0, 3, 3, 2, 1, 2, 5, 2, 5, 6, 2, 1, 8, 4, 7, 2, 7, 5, 7, 2, 7, 1, 8, 5, 5, 2, 8, 8, 6, 7, 3, 2, 1, 2, 4, 3, 3, 7, 6, 1, 0, 4, 3, 3, 1, 0, 7, 2, 2, 8, 1, 3, 2, 2, 3, 5, 0, 4, 1, 0, 3, 6, 8, 1, 4, 0, 7, 8, 7, 0, 8, 1, 3, 7, 4, 2, 8, 8, 5, 5, 3, 7, 6, 0, 1, 5, 7, 1, 8, 6, 2, 6, 4, 5, 1, 6] 
2.6.5 :070 > array.my_inject(:+)
Traceback (most recent call last):
        7: from /usr/share/rvm/rubies/ruby-2.6.5/bin/irb:23:in `<main>'
        6: from /usr/share/rvm/rubies/ruby-2.6.5/bin/irb:23:in `load'
        5: from /usr/share/rvm/rubies/ruby-2.6.5/lib/ruby/gems/2.6.0/gems/irb-1.0.0/exe/irb:11:in `<top (required)>'
        4: from (irb):70
        3: from enumerables.rb:114:in `my_inject'
        2: from enumerables.rb:12:in `my_each'
        1: from enumerables.rb:114:in `block in my_inject'
LocalJumpError (no block given (yield))
2.6.5 :071 > array.inject(:+)
 => 390 
2.6.5 :072 >
```

**Review by maelfosso — CHANGES_REQUESTED** [review]
# Changes required :red_circle: 

Hello,
Thanks for the applying the changes.
However, it's not sufficient. There are still some issues to fix

1. Kindly fix the stickler's errors, please.

2. Read the comments left below highlighting the errors

Make the suggested changes and submit a new code review request
Thanks

Mael FOSSO

**Review by MarioDena — CHANGES_REQUESTED** [review]
Hello team good work on implementing your classes however there are some changes needed.

## Changes required  🔴 

- [x] Please check all the comments by @maelfosso  After running the tests on your project all errors on #my_each and #my_inject still persist.

After making the changes please resubmit for another review.

If you need help in fixing the problems remember you can reach out student success by the dashboard and find help into understanding the problem further.

**Comment by fdfragoso** [comment]
@MarioDena The tests were made following ruby documentation and all the tests were working. Please check the live version with the tests of our code.

https://repl.it/@FelipeDacal/EnumerablesTestBench

**Comment by davidmrtz-dev** [comment]
## Hi Team. 👨‍🚀
### Very good job so far.


@fdfragoso To verify your project meets all the requirements as TSE we run a Rspec test suit that checks all your methods returns the correct value, testing the methods with multiple inputs and checking the outputs. 😄 

There are some issues with some methods that need to be implemented fixed, please check out the comments below.  👨‍💻  ⬇️ 

****
**Please make the changes required and submit for re-review once you are ready!** 
****
### Status: Changes required. ♻️  
****


**[TSE:  David Elí]** 
Slack: @ David Eli Martinez Garcia

File: enumerables.rb [hunk 15]
```diff
@@ -0,0 +1,135 @@
+# rubocop: disable Metrics/ModuleLength, Metrics/CyclomaticComplexity, Metrics/PerceivedComplexity
+module Enumerable
+  def my_each
```
Comment (davidmrtz-dev) [inline enumerables.rb:3]: Enumerable each method calls the given block once for each element in self, passing that element as a parameter. Returns the array itself.

Screen Shot 2020-06-27 at 2 45 36 PM

File: enumerables.rb [hunk 16]
```diff
@@ -0,0 +1,135 @@
...
+    return to_enum unless block_given?
+
+    i = 0
+    while i <= size - 1
+      yield(to_a[i])
+      i += 1
+    end
+  end
+
+  def my_each_with_index
```
Comment (davidmrtz-dev) [inline enumerables.rb:14]: Calls block with two arguments, the item and its index, for each item in enumerable.
When a block is given when self is an array returns the array itself after calling the given block once for each element in self, similar to each method that returns the array itself.

Screen Shot 2020-06-27 at 2 56 50 PM

File: enumerables.rb [hunk 17]
```diff
@@ -0,0 +1,135 @@
...
+        result = false if item
+      else
+        result = false if var === item
+      end
+    end
+    result
+  end
+
+  # rubocop: enable Style/CaseEquality, Style/IfInsideElse
+  def my_count(var = nil)
```
Comment (davidmrtz-dev) [inline enumerables.rb:79]: enumerable #my_count if a range is given returns the number of items in self.
Screen Shot 2020-06-27 at 3 04 14 PM

File: enumerables.rb (hunk 16 above)
Comment (fdfragoso) [reply enumerables.rb:14]: myeachindex

For what looks like the test is working. I didn't change anything. So I believe is working!

**Review by euqueme — CHANGES_REQUESTED** [review]
> ###  Pull Request under the review by Maria Eugenia Queme P. (TSE)

### Hello @fdfragoso and @luciano-ilha :lollipop: :balloon: 

### Thank you for working on the changes requested by previous reviewers. You've done an incredible job so far, :medal_sports: :medal_sports: almost all your enumerable methods are working properly showing no errors but there's a couple that still displays the wrong answers so next I'll do my best to explain the test cases that aren't passing with examples :point_down: 

#### Minimum Requirements

##### my_each_with_index

* [ ] should have working `my_each_with_index` identical and returns the same thing as ruby's [`each_with_index`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-each_with_index). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_each_with_index.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))

- [ ] my_each_with_index calls the given block once for each element in self

```ruby
#Given this block
block = proc { |x, i| x + i }

[1, 5, 2].my_each_with_index(&block)  == [1, 5, 2].each_with_index(&block) => [1, 5, 2]
```
- Yours returns nil at the end
image

- Real each_with_index returns the enumerable at the end
image

- [ ] my_each_with_index, when a block is given when self is a range, returns the range itself after calling the given block once for each element in self

```ruby
#Given this block
block = proc { |x, i| x + i }

(1..5).my_each_with_index(&block)  == (1..5).each_with_index(&block) => 1..5
```
- Yours returns nil at the end
image

- Real each_with_index returns the enumerable(range) at the end
image

#####  my_inject

- [ ] should have working `my_inject` identical and returns the same thing as ruby's [`inject`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-inject). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_inject.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))

- [ ] my_inject raises a "LocalJumpError" when no block or argument is given

```ruby
[1, 3, 4, 5].my_inject == [1, 3, 4, 5].inject => Should throw a LocalJumpError
```

- Yours throws no error and is returning the enumerable instead
image

- The real inject throws the expected error (correct answer)
image

### That's it, team! :raised_hands: 

### Please :pray: do the changes :technologist: :technologist: and submit back for a code review when you are done :memo: 

Happy coding! :rocket: 

Any question? ask me on slack **@ Maru** :woman_technologist: 

PROJECT STATUS: REQUESTED CHANGES :traffic_light:

**Comment by lekegitrepo** [comment]
#### Good job :clap: :clap:
#### This project is approved :rocket:

#### _Reviewed by Temitayo Adeleke_.

**Pull request closed without merging**